### PR TITLE
[libunwind][WebAssembly] Make libunwind compilable

### LINKED ISSUE #72771: How to build libunwind for wasm?
error: "Unsupported architecture."

```[14/16] Building CXX object src/CMakeFiles/unwind_static_objects.dir/libunwind.cpp.obj
FAILED: src/CMakeFiles/unwind_static_objects.dir/libunwind.cpp.obj
/home/cqwrteur/llvm_toolchains/x86_64-pc-linux-gnu/bin/clang++ --target=wasm32-wasi --sysroot=/home/cqwrteur/toolchains/sysroots/wasm32-wasi/wasm32-wasi -D_LIBCPP_HAS_NO_PRAGMA_SYSTEM_HEADER -D_LIBUNWIND_IS_BAREMETAL -I/home/cqwrteur/projects/llvm-project/libunwind/include -Wno-array-bounds -Wno-cast-function-type -Wno-uninitialized -fuse-ld=lld -Wno-misleading-indentation -Wno-global-constructors -Wno-unused-command-line-argument -O3 -DNDEBUG -std=c++11 -Werror=return-type -funwind-tables -nostdinc++ -D_DEBUG -UNDEBUG -D_LIBUNWIND_IS_NATIVE_ONLY -Wall -Wextra -Wnewline-eof -Wshadow -Wwrite-strings -Wno-unused-parameter -Wno-long-long -Werror=return-type -Wextra-semi -Wundef -Wunused-template -Wformat-nonliteral -Wno-user-defined-literals -Wno-covered-switch-default -Wno-suggest-override -Wno-error -pedantic -fno-rtti  -fstrict-aliasing -fno-exceptions -fno-rtti -MD -MT src/CMakeFiles/unwind_static_objects.dir/libunwind.cpp.obj -MF src/CMakeFiles/unwind_static_objects.dir/libunwind.cpp.obj.d -o src/CMakeFiles/unwind_static_objects.dir/libunwind.cpp.obj -c /home/cqwrteur/projects/llvm-project/libunwind/src/libunwind.cpp
In file included from /home/cqwrteur/projects/llvm-project/libunwind/src/libunwind.cpp:12:
In file included from /home/cqwrteur/projects/llvm-project/libunwind/include/libunwind.h:16:
/home/cqwrteur/projects/llvm-project/libunwind/include/__libunwind_config.h:184:4: error: "Unsupported architecture."
  184 | #  error "Unsupported architecture."
      |    ^
In file included from /home/cqwrteur/projects/llvm-project/libunwind/src/libunwind.cpp:12:
/home/cqwrteur/projects/llvm-project/libunwind/include/libunwind.h:71:17: error: use of undeclared identifier '_LIBUNWIND_CONTEXT_SIZE'
   71 |   uint64_t data[_LIBUNWIND_CONTEXT_SIZE];
      |                 ^
/home/cqwrteur/projects/llvm-project/libunwind/include/libunwind.h:76:17: error: use of undeclared identifier '_LIBUNWIND_CURSOR_SIZE'
   76 |   uint64_t data[_LIBUNWIND_CURSOR_SIZE];
      |                 ^
In file included from /home/cqwrteur/projects/llvm-project/libunwind/src/libunwind.cpp:30:
In file included from /home/cqwrteur/projects/llvm-project/libunwind/src/AddressSpace.hpp:23:
In file included from /home/cqwrteur/projects/llvm-project/libunwind/src/EHHeaderParser.hpp:17:
/home/cqwrteur/projects/llvm-project/libunwind/src/DwarfParser.hpp:69:26: error: use of undeclared identifier '_LIBUNWIND_HIGHEST_DWARF_REGISTER'
   69 |     kMaxRegisterNumber = _LIBUNWIND_HIGHEST_DWARF_REGISTER
      |                          ^
In file included from /home/cqwrteur/projects/llvm-project/libunwind/src/libunwind.cpp:31:
In file included from /home/cqwrteur/projects/llvm-project/libunwind/src/UnwindCursor.hpp:50:
/home/cqwrteur/projects/llvm-project/libunwind/src/RWMutex.hpp:19:10: fatal error: 'pthread.h' file not found
   19 | #include <pthread.h>
      |          ^~~~~~~~~~~
5 errors generated.
ninja: build stopped: subcommand failed.

### PR DESCRIPTION
**Pull request by aheejin**
This tries to make Wasm compilable in LLVM tree with CMake for non-Emscripten platform.

This
- Adds `-D__USING_WASM_EXCEPTIONS__` when you compile with `-fwasm-exceptions` (like other EH options) in Clang
- Exclude `UnwindLevel1.c`, `UnwindRegistersSave.S`, and `UnwindRegistersRestore.S` when compiling with Wasm
- Changed some `__USING_WASM_EXCEPTIONS__` to `__wasm__`; they should be applied when compiling with Wasm w/o exceptions.
- Define some unused macros to make it compile

Fixes #72771.

### CODE REVIEW

**Comment by aheejin** [comment]
Can someone review this?

File: libunwind/src/UnwindRegistersRestore.S [hunk 1]
```diff
@@ -20,7 +20,7 @@
   .text
 #endif
 
-#if !defined(__USING_SJLJ_EXCEPTIONS__)
+#if !defined(__USING_SJLJ_EXCEPTIONS__) && !defined(__wasm__)
```
Comment (asl) [inline libunwind/src/UnwindRegistersRestore.S:23]: You are defining `__USING_WASM_EXCEPTIONS__`. Why you are not using it here and everywhere else?

Comment (aheejin) [reply libunwind/src/UnwindRegistersRestore.S:23]: `__wasm__` is defined whenever we compile the code to wasm, and `__USING_WASM_EXCEPTIONS__` is defined when we compile the code to wasm and use Wasm exceptions (`-fwasm-exeptions`). And this code should be excluded whenever it is compiled to wasm, even if `-fwasm-exceptions` is not used.

Comment (asl) [reply libunwind/src/UnwindRegistersRestore.S:23]: Ok. Then does the clang change really belong here? `LIBUNWIND_USES_WASM_EXCEPTIONS` also seems to be unused.

Comment (aheejin) [reply libunwind/src/UnwindRegistersRestore.S:23]: Removed the clang change and `LIBUNWIND_USES_WASM_EXCEPTIONS`.

Comment (aheejin) [reply libunwind/src/UnwindRegistersRestore.S:23]: Can I land this? Thank you for your review!

**Comment by aheejin** [comment]
I'd appreciate if someone LGTM this. Thanks!

**Review by dschuff — APPROVED** [review]
These changes make sense from the wasm point of view, and it looks like you've addressed the comment, so LGTM

**Review by asl — APPROVED** [review]
Ok from the generic point of view. I cannot judge if wasm-related part is ok though :)

**Pull request merged**